# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/windows/ShellHeaderOnlyUtils.h
# Commit: cae98337ddda
# Full Hash: cae98337ddda1b78db120235072ef5154cdf65f5
# Author: Aaron Klotz <aklotz@mozilla.com>
# Date: 2019-07-24 09:54:28
# Regressor Bug: 1567614
# File Overlap Count: 1
# Description:
#   Bug 1567614: Part 2 - Add ShellExecuteByExplorer overload to handle absolute PIDL lists; r=jmathies
#   
#   For the URI handling case, we still want to parse the URI to look for any
#   malformation. OTOH, IShellDispatch2::ShellExecute does not accept PIDLs as
#   arguments, we we need an overload that converts the absolute PIDL back to a
# ==============================================================================

diff -r e8bf96e8321a -r cae98337ddda widget/windows/ShellHeaderOnlyUtils.h
--- a/widget/windows/ShellHeaderOnlyUtils.h	Tue Jul 23 20:17:58 2019 +0000
+++ b/widget/windows/ShellHeaderOnlyUtils.h	Tue Jul 23 20:18:07 2019 +0000
@@ -147,6 +147,62 @@
   return Ok();
 }
 
+using UniqueAbsolutePidl =
+    UniquePtr<RemovePointer<PIDLIST_ABSOLUTE>::Type, CoTaskMemFreeDeleter>;
+
+inline UniqueAbsolutePidl ShellParseDisplayName(const wchar_t* aPath) {
+  PIDLIST_ABSOLUTE rawAbsPidl = nullptr;
+  SFGAOF sfgao;
+  HRESULT hr = ::SHParseDisplayName(aPath, nullptr, &rawAbsPidl, 0, &sfgao);
+  if (FAILED(hr)) {
+    return nullptr;
+  }
+
+  return UniqueAbsolutePidl(rawAbsPidl);
+}
+
+/**
+ * Since IShellDispatch2::ShellExecute does not accept a PIDL as one of its
+ * parameters, we need to convert a PIDL back to a path. This overload handles
+ * that conversion and then proceeds with the ShellExecuteByExplorer call.
+ */
+inline LauncherVoidResult ShellExecuteByExplorer(
+    const UniqueAbsolutePidl& aPath, const _variant_t& aArgs,
+    const _variant_t& aVerb, const _variant_t& aWorkingDir,
+    const _variant_t& aShowCmd) {
+  // |aPath| is an absolute path. IShellFolder::GetDisplayNameOf requires a
+  // valid child ID, so the first thing we need to resolve is the IShellFolder
+  // for |aPath|'s parent, as well as the childId that represents |aPath|.
+  // Fortunately SHBindToParent does exactly that!
+  PCUITEMID_CHILD childId = nullptr;
+  RefPtr<IShellFolder> parentFolder;
+  HRESULT hr = ::SHBindToParent(aPath.get(), IID_IShellFolder,
+                                getter_AddRefs(parentFolder), &childId);
+  if (FAILED(hr)) {
+    return LAUNCHER_ERROR_FROM_HRESULT(hr);
+  }
+
+  // Now we retrieve the display name of |childId|, telling the shell that we
+  // plan to have the string parsed.
+  STRRET strret;
+  hr = parentFolder->GetDisplayNameOf(childId, SHGDN_FORPARSING, &strret);
+  if (FAILED(hr)) {
+    return LAUNCHER_ERROR_FROM_HRESULT(hr);
+  }
+
+  // Since ShellExecuteByExplorer wants a BSTR, we convert |strret|. Calling
+  // StrRetToBSTR is advantageous since it automatically takes care of
+  // freeing any dynamically allocated memory in |strret|.
+  _bstr_t bstrPath;
+  hr = ::StrRetToBSTR(&strret, nullptr, bstrPath.GetAddress());
+  if (FAILED(hr)) {
+    return LAUNCHER_ERROR_FROM_HRESULT(hr);
+  }
+
+  // Now we have a bstr_t so we can invoke the overload.
+  return ShellExecuteByExplorer(bstrPath, aArgs, aVerb, aWorkingDir, aShowCmd);
+}
+
 }  // namespace mozilla
 
 #endif  // mozilla_ShellHeaderOnlyUtils_h